odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+        ...                  [3, 4, 5]])
+        >>> x in spc
+        True
+
+        Sizes, data types and other essential properties characterize
+        spaces and decide about membership:
+
+        >>> smaller_spc = odl.tensor_space((2, 2), dtype='uint64')
+        >>> y = smaller_spc.element([[0, 1],
+        ...                          [2, 3]])
+        >>> y in spc
+        False
+        >>> x in smaller_spc
+        False
+        >>> other_dtype_spc = odl.tensor_space((2, 3), dtype='uint32')
+        >>> z = other_dtype_spc.element([[0, 1, 2],
+        ...                              [3, 4, 5]])
+        >>> z in spc
+        False
+        >>> x in other_dtype_spc
+        False
+
+        On the other hand, spaces are not unique:
+
+        >>> spc2 = odl.tensor_space((2, 3), dtype='uint64')  # same as spc
+        >>> x2 = spc2.element([[5, 4, 3],
+        ...                    [2, 1, 0]])
+        >>> x2 in spc
+        True
+        >>> x in spc2
+        True
+
+        Of course, random garbage is not in the space:
+
+        >>> spc = odl.tensor_space((2, 3), dtype='uint64')
+        >>> None in spc
+        False
+        >>> object in spc
+        False
+        >>> False in spc
+        False
+        """
+        return getattr(other, 'space', None) == self
+
+    def __eq__(self, other):
+        """Return ``self == other``.
+
+        Returns
+        -------
+        equals : bool
+            True if ``self`` and ``other`` have the same type, `shape`
+            `dtype` and `order`, otherwise ``False``.
+
+        Examples
+        --------
+        Sizes, data types and other essential properties characterize
+        spaces and decide about equality:
+
+        >>> spc = odl.tensor_space(3, dtype='uint64')
+        >>> spc == spc
+        True
+        >>> spc2 = odl.tensor_space(3, dtype='uint64')
+        >>> spc2 == spc
+        True
+        >>> smaller_spc = odl.tensor_space(2, dtype='uint64')
+        >>> spc == smaller_spc
+        False
+        >>> other_dtype_spc = odl.tensor_space(3, dtype='uint32')
+        >>> spc == other_dtype_spc
+        False
+        >>> other_shape_spc = odl.tensor_space((3, 1), dtype='uint64')
+        >>> spc == other_shape_spc
+        False
+        """
+        if other is self:
+            return True
+
+        return (type(self) == type(other) and
+                self.shape == other.shape and
+                self.dtype == other.dtype and
+                self.order == other.order)
+
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((type(self), self.shape, self.dtype, self.order))
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        posargs = [self.shape, dtype_str(self.dtype)]
+        optargs = [('order', self.order, 'A')]
+        return "{}({})".format(self.__class__.__name__,
+                               signature_string(posargs, optargs))
+
+    __str__ = __repr__
```

Comment:
That's actually true, somewhat surprising to me (both assignment and override at class level, so no clear order). Should be changed then.